transport/grpc/handler.go

```diff
@@ -158,17 +119,100 @@ func procedureFromStreamMethod(streamMethod string) (string, error) {
 	return procedureToName(service, method)
 }
 
-func (h *handler) call(ctx context.Context, transportRequest *transport.Request, responseWriter *responseWriter) error {
-	handlerSpec, err := h.i.router.Choose(ctx, transportRequest)
-	if err != nil {
+func (h *handler) handleStream(
+	ctx context.Context,
+	transportRequest *transport.Request,
+	serverStream grpc.ServerStream,
+	start time.Time,
+	streamHandler transport.StreamHandler,
+) error {
+	tracer := h.i.t.options.tracer
+	if tracer == nil {
+		tracer = opentracing.GlobalTracer()
+	}
+	var parentSpanCtx opentracing.SpanContext
+	md, ok := metadata.FromIncomingContext(ctx)
+	if ok {
+		parentSpanCtx, _ = tracer.Extract(opentracing.HTTPHeaders, mdReadWriter(md))
+	}
+	extractOpenTracingSpan := &transport.ExtractOpenTracingSpan{
+		ParentSpanContext: parentSpanCtx,
+		Tracer:            tracer,
+		TransportName:     transportName,
+		StartTime:         start,
+	}
+	ctx, span := extractOpenTracingSpan.Do(ctx, transportRequest)
+	defer span.Finish()
+
+	return transport.UpdateSpanWithErr(span, transport.DispatchStreamHandler(
+		streamHandler,
+		newServerStream(ctx, transportRequest.ToRequestMeta(), serverStream),
+	))
+}
+
+func (h *handler) handleUnary(
+	ctx context.Context,
+	transportRequest *transport.Request,
+	serverStream grpc.ServerStream,
+	streamMethod string,
+	start time.Time,
+	handler transport.UnaryHandler,
+) error {
+	var requestData []byte
+	if err := serverStream.RecvMsg(&requestData); err != nil {
 		return err
 	}
-	switch handlerSpec.Type() {
-	case transport.Unary:
-		return h.callUnary(ctx, transportRequest, handlerSpec.Unary(), responseWriter)
-	default:
-		return yarpcerrors.Newf(yarpcerrors.CodeUnimplemented, "transport grpc does not handle %s handlers", handlerSpec.Type().String())
+	requestBuffer := bufferpool.Get()
+	defer bufferpool.Put(requestBuffer)
+
+	// Buffers are documented to always return a nil error.
+	_, _ = requestBuffer.Write(requestData)
+	transportRequest.Body = requestBuffer
+
+	responseWriter := newResponseWriter()
+	defer responseWriter.Close()
+
+	err := h.handleUnaryBeforeErrorConversion(ctx, transportRequest, responseWriter, start, handler)
+	err = handlerErrorToGRPCError(err, responseWriter)
+
+	// Send the response attributes back and end the stream.
+	if sendErr := serverStream.SendMsg(responseWriter.Bytes()); sendErr != nil {
+		// We couldn't send the response.
+		return sendErr
+	}
+	if responseWriter.md != nil {
+		serverStream.SetTrailer(responseWriter.md)
+	}
+	return err
+}
+
+func (h *handler) handleUnaryBeforeErrorConversion(
+	ctx context.Context,
+	transportRequest *transport.Request,
+	responseWriter *responseWriter,
+	start time.Time,
+	handler transport.UnaryHandler,
+) error {
+	tracer := h.i.t.options.tracer
+	if tracer == nil {
+		tracer = opentracing.GlobalTracer()
+	}
+	var parentSpanCtx opentracing.SpanContext
```

Comment:
The opentracing logic - could this be pulled into a private function below that is called by both the unary and stream paths?

Reply:
Would be nice.